visualization/include/pcl/visualization/impl/pcl_visualizer.hpp

```diff
@@ -715,6 +715,71 @@ pcl::visualization::PCLVisualizer::addText3D (
   return (true);
 }
 
+//////////////////////////////////////////////////
+template <typename PointT> bool
+pcl::visualization::PCLVisualizer::addText3D (
+  const std::string &text,
+  const PointT& position,
+  double orientation[3],
+  double textScale,
+  double r,
+  double g,
+  double b,
+  const std::string &id,
+  int viewport)
+{
+  std::string tid;
+  if (id.empty ())
+    tid = text;
+  else
+    tid = id;
+
+  if (contains (tid))
+  {
+    PCL_WARN ("[addText3D] The id <%s> already exists! Please choose a different id and retry.\n", id.c_str ());
+    return (false);
+  }
+
+  vtkSmartPointer<vtkVectorText> textSource = vtkSmartPointer<vtkVectorText>::New ();
+  textSource->SetText (text.c_str());
+  textSource->Update ();
+
+  vtkSmartPointer<vtkPolyDataMapper> textMapper = vtkSmartPointer<vtkPolyDataMapper>::New ();
+  textMapper->SetInputConnection (textSource->GetOutputPort ());
+
+  // Since each follower may follow a different camera, we need different followers
+  rens_->InitTraversal ();
+  vtkRenderer* renderer = NULL;
+  int i = 0;
+  while ((renderer = rens_->GetNextItem ()) != NULL)
+  {
+    // Should we add the actor to all renderers or just to i-nth renderer?
+    if (viewport == 0 || viewport == i)
+    {
+      vtkSmartPointer<vtkActor> textActor = vtkSmartPointer<vtkActor>::New ();
```

Comment:
I think I would create a third method in which you supply the vtkActor in the signature, which is invoked by both methods. That way you can specialize the vtkActor for both use cases and the rest is the same. The only thing I don't like is exposing vtk types in a function signature, but I don't think I'm able to hide it from the public API because the methods which would call it are templated on the point type.

Reply:
If you want, I can try do an attempt in this way. As you say, it explose VTK types... Maybe you can hide this if you propose in the API 2 methods:
 - `addText3D_follow()`: follow the view point of the camera
 - `addText3D_oriented()`: the text stay oriented independently of the camera

You keep the `addText3D()` function for historical reason, and this function just call `addText3D_follow()`. Behind `addText3D_follow()` and `addText3D_oriented()` you manage the VTK types internally.

Reply:
It sounds to me that this is exactly the same situation as now, only that instead of making use of function overload, you're assigning different names to functions.

@taketwo what's your feeling on exposing an overload which allows you to provide the vtkActor directly, in the public API?

Reply:
I don't like exposing VTK classes in the API and would ideally view the fact that we use VTK backend as an implementation detail. (True, we have some VTK classes in the interface of `addModelFromPolyData()` family of functions, but that's it.)

Now getting back to this specific function, I don't get why we need to create multiple actors (one per viewport) in the oriented text case. Unlike the follow-camera text, they are all the same, aren't they?

Reply:
@taketwo : you should be right, so the `textActor` variable should be created outside the while loop, isn't it?

Reply:
I don't see the need for the loop at all. Just use `addActorToRenderer (actor, viewport)` as in other functions that add something to the visualizer.

Reply:
It's done.